### PR TITLE
Offer fix-its to disambiguate based on a trailing closure's label.

### PR DESCRIPTION
**Pull request by jrose-apple**
(by making it a normal argument with a label and not a trailing closure)

Diagnostic part of rdar://problem/25607552. A later commit will keep us from getting in this situation quite so much when default arguments are involved.

This PR also sneaks in a new frontend option for compiler hackers, -verify-apply-fixes, to enable @lattner's auto-apply-fixes mode for -verify. That's not really the focus here, and I don't plan on cherry-picking that to the 3.0 branch.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@nkcsgexi, @DougGregor, what do you think? @rintaro, are there any test cases I missed that came up when you were doing the rename fix-it?

**Comment by nkcsgexi** [comment]
Cool stuff! I wish we had this before so that the migrator does not need to worry about ambiguity after renaming.

**Comment by jrose-apple** [comment]
Yeah, unfortunately it happens *after* the migrator's already made its changes, so it can't just be auto-applied. But I figure you have good experience with these kinds of diagnostics by this point to be a good reviewer.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by nkcsgexi** [comment]
oh, You are right. We've no way to select which note's fixit to apply after wiping out the original code. I've briefly read the commit and it looks good to me.

**Comment by jrose-apple** [comment]
[Apple](https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx/3612/), [Linux](https://ci.swift.org/view/Pull%20Request/job/swift-PR-Linux/3771/)

**Comment by jrose-apple** [comment]
Added one more diagnostic cleanup: 71b2bd39fae831a2b68f3cb115f3f0cd8e102a21. Would someone look over that too?

**Comment by jrose-apple** [comment]
Previous tests passed, so I'm just going to smoke test the latest addition.

@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: test/expr/closure/trailing.swift [hunk 1]
```diff
@@ -180,3 +180,65 @@ func r23036383(foo: Foo23036383?, obj: Foo23036383) {
...
+func overloadOnDefaultArgsOnly3(x: Bool = true, a: () -> Void) {} // expected-note 2 {{found this candidate}}
+
+func overloadOnSomeDefaultArgsOnly(_: Int, x: Int = 0, a: () -> Void) {} // expected-note {{found this candidate}}
+func overloadOnSomeDefaultArgsOnly(_: Int, y: Int = 1, a: () -> Void) {} // expected-note {{found this candidate}}
+
+func overloadOnSomeDefaultArgsOnly2(_: Int, x: Int = 0, a: () -> Void) {} // expected-note {{found this candidate}}
+func overloadOnSomeDefaultArgsOnly2(_: Int, y: Bool = true, a: () -> Void) {} // expected-note {{found this candidate}}
+
+func overloadOnSomeDefaultArgsOnly3(_: Int, x: Int = 0, a: () -> Void) {} // expected-note {{found this candidate}}
+func overloadOnSomeDefaultArgsOnly3(_: Int, x: Bool = true, a: () -> Void) {} // expected-note {{found this candidate}}
```
Comment (rintaro) [inline test/expr/closure/trailing.swift]: Fixing
```swift
overloadOnLabelSomeDefaultArgs(
  1, x: 2
) {
  // some
}
```
I prefer
```swift
overloadOnLabelSomeDefaultArgs(
  1, x: 2, a: {
  // some
})
```
than
```swift
overloadOnLabelSomeDefaultArgs(
  1, x: 2
, a: {
  // some
})
```

Comment (jrose-apple) [inline test/expr/closure/trailing.swift]: That's a good point. Should be doable. (And at some point it's probably worth factoring out "call-related source location utilities", but I'm hoping to cherry-pick this and want to keep it scoped.)

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Note: split into a separate Radar, rdar://problem/28325311, so that rdar://problem/25607552 can deal with the fact that many such calls shouldn't be ambiguous in the first place (because one of the overloads has default arguments and the other doesn't).

**Pull request merged**